trl/models/modeling_vhead.py

```diff
@@ -0,0 +1,113 @@
+# Copyright 2022 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import torch.nn as nn
+from transformers import AutoModelForCausalLM
+
+from .modeling_base import PreTrainedModelWrapper
+
+class ValueHead(nn.Module):
+    r"""
+    The ValueHead class implements a head for GPT2 that returns a scalar for each output token.
+    """
+    def __init__(self, config):
+        super().__init__()
+        self.detach_head = False
+        self.summary_type = config.summary_type if hasattr(config, "summary_type") else "last"
+        if self.summary_type == "attn":
+            raise NotImplementedError
+
+        self.summary = nn.Identity()
+        if hasattr(config, "summary_use_proj") and config.summary_use_proj:
+            if hasattr(config, "summary_proj_to_labels") and config.summary_proj_to_labels and config.num_labels > 0:
+                num_classes = config.num_labels
+            else:
+                num_classes = config.hidden_size
+            self.summary = nn.Linear(config.hidden_size, num_classes)
+
+        self.activation = nn.Identity()
+        if hasattr(config, "summary_activation") and config.summary_activation == "tanh":
+            self.activation = nn.Tanh()
+
+        self.first_dropout = nn.Identity()
+        if hasattr(config, "summary_first_dropout") and config.summary_first_dropout > 0:
+            self.first_dropout = nn.Dropout(config.summary_first_dropout)
+
+        self.last_dropout = nn.Identity()
+        if hasattr(config, "summary_last_dropout") and config.summary_last_dropout > 0:
+            self.last_dropout = nn.Dropout(config.summary_last_dropout)
```

Comment:
This needs to be simplified in the future

Reply:
From what I can see, we'll probably need an extra arg (`v_head_activation`) and simplify this, how does that sound @edbeeching @lvwerra ?

Reply:
I guess the `v_head` never has an activation but maybe @edbeeching knows better so activation is actually not need at all. Also not clear if we ever want to apply dropout to the output (`last_dropout`).

Reply:
also all the summary attributes are not needed as far as I can tell. maybe all we need is dropout and then a linear layer. can you check what the defaults are?

Reply:
Here are the defaults: 
```
(v_head): ValueHead(
    (summary): Linear(in_features=768, out_features=2, bias=True)
    (activation): Identity()
    (first_dropout): Dropout(p=0.1, inplace=False)
    (last_dropout): Identity()
    (flatten): Flatten(start_dim=1, end_dim=-1)
  )
```
Let's stick to this architecture now and address a potential adding of activation etc in a follow up PR

Reply:
also it seems that `flatten` is never used, I removed it

Reply:
I have never seen a vhead with an activation. You can imagine situations that might use one, but they are so rare that I agree this is not needed.

Reply:
Perfect thank you for confirming !!